clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1011,8 +996,14 @@ private Record nextFetchedRecord() {
                         drain();
                         return null;
                     }
-                    currentBatch = batches.next();
-                    maybeEnsureValid(currentBatch);
+
+                    if (hasExceptionInLastFetch) {
+                        currentBatch.ensureValid();
+                    } else {
+                        lastRecord = null;
+                        currentBatch = batches.next();
+                        maybeEnsureValid(currentBatch);
```

Comment:
If a batch validation failed, next time when it comes into the loop it will go to line 1001 and hit the same validation error again right? So we will not continue to record validation. But it is true that we should do that before line 995. I'll update patch to fix that.